### PR TITLE
Sbom

### LINKED ISSUE #1615: `pack sbom download` should be able to retrieve the SBoM from remote images
### Description

When a user tries to download an `SBoM` from an image BUT the image is not present in the docker daemon, pack throws an error similar to `ERROR: image 'image' cannot be found: not found`. The user is forced to download the image to the docker daemon to be able to extract the SBoM information. 

This behavior is a little weird, but I am not sure if it is because some technical limitation or pack could retrieved the SBoM from the registry.

### Proposed solution

1. Investigate why the SBoM is only retrieved from the docker daemon
2. In case the SBoM can be retrieved from the registry then download the file
3. If the image can't be found in a registry or in the daemon, then show the `can not be found` error

### Describe alternatives you've considered

No other alternatives considered

### Additional context
- No documentation required
- This issue is related to #1614

### PR DESCRIPTION
**Pull request by bhavyastar**
## Summary
It changes the message in download_sbom.go as we can download sbom from the remote registry with the help of `--remote ` flag

- Should this change be documented?
    - [ ] Yes, see #___
    - [x] No

Resolves #1615

### CODE REVIEW

**Comment by bhavyastar** [comment]
CC  @natalieparellano @jjbustamante

File: pkg/client/download_sbom.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ func (c *Client) DownloadSBOM(name string, options DownloadSBOMOptions) error {
 	img, err := c.imageFetcher.Fetch(context.Background(), name, image.FetchOptions{Daemon: options.Daemon, PullPolicy: image.PullNever})
 	if err != nil {
 		if errors.Cause(err) == image.ErrNotFound {
-			c.logger.Warnf("if image is saved on a registry run 'docker pull %s' before downloading the SBoM", name)
+			c.logger.Warnf("if image is saved on a registry run with flag '--remote %s' before downloading the SBoM", name)
```
Comment (jjbustamante) [inline pkg/client/download_sbom.go:42]: Maybe something like:

"if the image is saved on a registry run with the flag '--remote', for example: ' pack sbom download --remote %s'" ?

Comment (bhavyastar) [reply pkg/client/download_sbom.go]: Yes!! Corrected!! @jjbustamante

**Comment by bhavyastar** [comment]
Hey @jjbustamante, Now, I have updated the unit test file as well. Can you check now if is it the right way to do this?

**Comment by bhavyastar** [comment]
If it is not the right way, I wanna learn more about this and do contribute in future as well!! Thanks for helping me out!! @jjbustamante

**Comment by jjbustamante** [comment]
> If it is not the right way, I wanna learn more about this and do contribute in future as well!! Thanks for helping me out!! @jjbustamante

Hi @bhavyastar! thanks for keep working on this :), it is still not ready, but don't feel bad about it. Let me try to guide you a little bit.

1. First of all, I don't know why the CI on github is not failing the unit test. I used a [tool called gswt](https://github.com/aemengo/gswt) created by a former member of the team to check the log output from your pull request. As we can see in the image, your unit test is still failing 😞 

Screenshot 2023-03-01 at 9 19 25 AM

2. The first think you need to do everytime before pushing your code to the repo, not just here on this contribution, but as a good practice on software development is to run the unit test suite in your local environment to verify that everything is fine after any changed you did on the base code. on pack you can do that running `make test command`. see development [guide](https://github.com/buildpacks/pack/blob/main/DEVELOPMENT.md#testing)

3. You can also use your IDE to run the test in a more user friendly UI, but it's up to each developer, in any case, for your particular issue, you can also try using this command in your terminal.

This is the output from my terminal where I pulled your branch.

```bash
➜  pack git:(sbom) go test -run ^TestDownloadSBOM$ github.com/buildpacks/pack/pkg/client
Suite: DownloadSBOM
Total: 5 | Focused: 0 | Pending: 0
...x.
Passed: 4 | Failed: 1 | Skipped: 0

--- FAIL: TestDownloadSBOM (0.00s)
    --- FAIL: TestDownloadSBOM/DownloadSBOM (0.00s)
        --- FAIL: TestDownloadSBOM/DownloadSBOM/the_image_doesn't_exist/returns_nil (0.00s)
            download_sbom_test.go:117: Expected 'Warning: if the image is saved on a registry run with the flag '--remote', for example: ' pack sbom download --remote some/non-existent-image' before downloading the SBoM
                ' to contain 'if image is saved on a registry run with the flag '--remote', for example: 'pack sbom download --remote %s' before downloading the SBoM
                '

                Diff:  strings.Join({
                - 	"if",
                + 	"Warning: if the",
                  	" image is saved on a registry run with the flag '--remote', for ",
                  	"example: '",
                + 	" ",
                  	"pack sbom download --remote ",
                - 	"%s",
                + 	"some/non-existent-image",
                  	"' before downloading the SBoM\n",
                  }, "")
FAIL
FAIL	github.com/buildpacks/pack/pkg/client	0.517s
FAIL
```

Again, we can see the test is still failing 😞 

4. In order to fix your test, you need to assert the expected output message of the method you are testing using the parameters. You are calling the method as follow: 

```go
subject.DownloadSBOM("some/non-existent-image", DownloadSBOMOptions{Daemon: true, DestinationDir: ""})
```
your image name is `some/non-existent-image`, you need to format the expected error message (you could use fmt.Sprintf) with that value and then do the Assert. In your local machine, just follow those steps, update your code and run the test locally until you fix the error. 

Hope this can be helpful, and again, thanks a lot for your contributions! ❤️

File: pkg/client/download_sbom.go [hunk 2]
```diff
@@ -39,7 +39,7 @@ func (c *Client) DownloadSBOM(name string, options DownloadSBOMOptions) error {
 	img, err := c.imageFetcher.Fetch(context.Background(), name, image.FetchOptions{Daemon: options.Daemon, PullPolicy: image.PullNever})
 	if err != nil {
 		if errors.Cause(err) == image.ErrNotFound {
-			c.logger.Warnf("if image is saved on a registry run 'docker pull %s' before downloading the SBoM", name)
+			c.logger.Warnf("if the image is saved on a registry run with the flag '--remote', for example: ' pack sbom download --remote %s' before downloading the SBoM", name)
```
Comment (natalieparellano) [inline pkg/client/download_sbom.go:42]:
```suggestion
			c.logger.Warnf("if the image is saved on a registry run with the flag '--remote', for example: ' pack sbom download --remote %s'", name)
```

File: pkg/client/download_sbom_test.go [hunk 3]
```diff
@@ -114,7 +114,7 @@ func testDownloadSBOM(t *testing.T, when spec.G, it spec.S) {
 
 			err := subject.DownloadSBOM("some/non-existent-image", DownloadSBOMOptions{Daemon: true, DestinationDir: ""})
 			h.AssertError(t, err, "image 'some/non-existent-image' cannot be found")
-			h.AssertContains(t, out.String(), "if image is saved on a registry run 'docker pull some/non-existent-image' before downloading the SBoM\n")
+			h.AssertContains(t, out.String(), "if image is saved on a registry run with the flag '--remote', for example: 'pack sbom download --remote %s' before downloading the SBoM\n")
```
Comment (natalieparellano) [inline pkg/client/download_sbom_test.go:117]:
```suggestion
			h.AssertContains(t, out.String(), "if the image is saved on a registry run with the flag '--remote', for example: 'pack sbom download --remote %s'\n")
```

**Comment by natalieparellano** [comment]
@bhavyastar thanks for this! This warning will really come in handy :)

I think if you accept my code suggestion, your unit test should pass. The other failing tests aren't due to your PR - when https://github.com/buildpacks/pack/pull/1657 lands we can pull in `main` again, then this should be ready to merge.

**Comment by bhavyastar** [comment]
It is still failing! :(  @natalieparellano and @jjbustamante . What should I do now?

**Comment by jjbustamante** [comment]
> your image name is `some/non-existent-image`, you need to format the expected error message (you could use fmt.Sprintf) with that value and then do the Assert. In your local machine, just follow those steps, update your code and run the test locally until you fix the error.

Did you try to do this? the test is failing because you are not formatting the expected message with the image name

**Comment by natalieparellano** [comment]
Whoops, I missed that part @jjbustamante. `fmt.Sprintf` should do the trick :)

**Comment by bhavyastar** [comment]
```
when("the image doesn't exist", func() {
		it("returns nil", func() {
			mockImageFetcher.EXPECT().Fetch(gomock.Any(), "some/non-existent-image", image.FetchOptions{Daemon: true, PullPolicy: image.PullNever}).Return(nil, image.ErrNotFound)
	
			err := subject.DownloadSBOM("some/non-existent-image", DownloadSBOMOptions{Daemon: true, DestinationDir: ""})
			expectedError := fmt.Sprintf("image '%s' cannot be found", "some/non-existent-image")
			h.AssertError(t, err, expectedError)
	
			expectedMessage := fmt.Sprintf("if the image is saved on a registry run with the flag '--remote', for example: 'pack sbom download --remote %s'", "some/non-existent-image")
			h.AssertContains(t, out.String(), expectedMessage)
		})
	})
```

I have written this code now but it is still showing that the test case is failing! 
Is it the right way to do it?
CC: @jjbustamante ( I guess, I am too curious and a keen learner that's why I asked so many questions. Sorry for disturbing you again and again)

**Comment by jjbustamante** [comment]
There two things that are making your test to fail:
1. 
```go 
expectedMessage := fmt.Sprintf("if the image is saved on a registry run with the flag '--remote', for example: 'pack sbom download --remote %s'", "some/non-existent-image")
``` 
Your output message is a `logger.Warnf` that means it is appending a 'Warning: ' in the log, you are missing that part in your `expectedMessaged` variable

2. The second one, is whitespace in your message at `download_sbom.go`.
```go
for example: ' pack sbom download
```
see the whitespace before pack?

**Comment by bhavyastar** [comment]
Okay, thanks @jjbustamante. Thanks a lot but this is the right way to do like what you said, I guess after this we are good to go!!

**Comment by bhavyastar** [comment]
Did some changes @jjbustamante! I hope this time it works!!

**Comment by jjbustamante** [comment]
> Did some changes @jjbustamante! I hope this time it works!!

Hi @bhavyastar, it failed formatting the code 😢 , could you run: `make format` locally and then push your changes again?

**Review by jkutner — APPROVED** [review]

**Comment by bhavyastar** [comment]
@jkutner Hey, Is this ready to merge?

**Comment by natalieparellano** [comment]
@bhavyastar looks like it's still failing on format. I think if you run `make format` and push the changes as suggested by @jjbustamante, it should go green :)

**Comment by jjbustamante** [comment]
Hi @bhavyastar could you update the branch and run `make format`?

**Comment by jjbustamante** [comment]
@bhavyastar Could you run `make format` and push again your changes?

**Comment by jjbustamante** [comment]
Closing this one because the issue was already fixed with [PR 2020](https://github.com/buildpacks/pack/pull/2020)

**Pull request closed without merging**